### PR TITLE
Adds CI hooks for GLibc compatibility check

### PR DESCRIPTION
**Pull request by tcsc**
Adds the GCB build yaml for controlling the build, and updates the test script to work in both the GCB environment and on a local dev machine.

Also changes the centos buildbox to leave the default user as `root`.  When GCB mounts the workspace into the container, the source code is owned by `root`, and there is no way to change this. This means that the build will fail when the non-root user specified in the build image attempt to write files into the workspace. Setting the root user fixes this.

See-Also: #15186

### CODE REVIEW

File: build.assets/build-test-compat.sh [hunk 1]
```diff
@@ -51,25 +51,54 @@ DISTROS=(
 # It will be set to non-zero value if any of run commands returns an error.
 EXIT_CODE=0
 
-# Run binary in a Docker container and propagate returned exit code.
+# Run binary in a Docker container and propagate returned exit code. 
+#
+# This will sometimes run under Google Cloud Build, which implies using Docker-
+# out-of-Docker to interact with containers. This means that simply mounting
+# the test targest into the test container won't work, as it would require
```
Comment (jakule) [inline build.assets/build-test-compat.sh:58]: Are you sure that mounting doesn't work? https://github.com/GoogleCloudPlatform/cloud-builders/issues/372#issuecomment-423692532

**Review by jakule — CHANGES_REQUESTED** [review]

File: .cloudbuild/ci/os-compatibility-test.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    entrypoint: "/bin/bash"
+    args: 
+      - '-c'
+      - 'make -j$$(nproc) build/tctl build/tsh build/tbot build/teleport'
+    timeout: 10m
+  - name: gcr.io/cloud-builders/docker
+    id: run-tests
+    entrypoint: "/bin/bash"
+    args: 
+      - './build.assets/build-test-compat.sh'
```
[the remark this answers is not in the document]
Comment (tcsc) [reply .cloudbuild/ci/os-compatibility-test.yaml:19]: The answer is "yes and no". The `docker` builder image _does_ have `make`, but it doesn't have `go` or a few other things our makefile needs. Running `make test-compat` still runs, but issues a heap of warnings that will probably just confuse people looking at test results.

File: build.assets/build-test-compat.sh (hunk 1 above)
Comment (tcsc) [reply build.assets/build-test-compat.sh:58]: That works for mounting stuff into the build container, but _not_ for mounting into the nested (well, actually _sibling_) containers we use for testing.

The Docker daemon we're talking to in this script is _outside_ the container that this script is running in. If we try to mount a directory into the child container then path that we give will be interpreted _in the context of that external Docker daemon_ and not our current container. We have no knowledge of how the build container is set up, so we can't translate our local path into one that the Docker daemon will interpret correctly.

If we had more control over the build container creation we could get expose a volume on it and use `docker run --volumes-from` pick it up. Unfortunately the build container is created by GCB and there are no options to _expose_ a volume, only to mount them, and so this is also a non starter.

File: .cloudbuild/ci/os-compatibility-test.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+timeout: 25m
+
+options:
+   machineType: E2_HIGHCPU_32
+
+steps:
+  - name: quay.io/gravitational/teleport-buildbox:teleport11
```
[the remark this answers is not in the document]
Comment (tcsc) [reply .cloudbuild/ci/os-compatibility-test.yaml:7]: To be honest, we should change _all_ of our CI to use the same environment we ship from. I can't imagine there are a huge variety of behavioral changes from using a slightly different glibc - but if we're going to run tests we should do it as close to the shipping config as possible.

[the remark this answers is not in the document]
Comment (jakule) [reply .cloudbuild/ci/os-compatibility-test.yaml]: I agree, for the most part, but I'm not sure if all the applications that we use in our CI runs on CentOS 7. I know that some apps stopped supporting CenOS 7 as the kernel and libraries were too old. Kernel shouldn't be a problem in our case, but switching all our CI images sounds like a very time-consuming task.

**Review by jakule — APPROVED** [review]

File: .cloudbuild/ci/os-compatibility-test.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+   machineType: E2_HIGHCPU_32
+
+steps:
+  - name: public.ecr.aws/gravitational/teleport-buildbox-centos7:teleport11
+    id: build-teleport
+    dir: /workspace
+    entrypoint: "/bin/bash"
+    args: 
+      - '-c'
+      - 'make -j$$(nproc) build/tctl build/tsh build/tbot build/teleport'
```
Comment (zmb3) [inline .cloudbuild/ci/os-compatibility-test.yaml:13]:
```suggestion
      - 'make -j$$(nproc) all'  
```

Won't this have the same effect? If so, it's less typing and we don't have to update it if we add a binary.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**